Public/Src/Cache/ContentStore/Library/Stores/ContentHashAddressableSnapshot.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using BuildXL.Cache.ContentStore.Hashing;
+
+namespace BuildXL.Cache.ContentStore.Stores
+{
+    /// <summary>
+    /// This class is specifically for usage within <see cref="ContentHashAddressableSnapshot{T}"/>. It is a reference type on purpose, in order to avoid issues
+    /// with the maximum object size limit.
+    /// </summary>
+    /// <typeparam name="T">Type tagged with a hash</typeparam>
+    public class PayloadFromDisk<T> : IComparable<PayloadFromDisk<T>>
+    {
+        /// <summary>
+        /// Hash for the <see cref="Payload"/>
+        /// </summary>
+        public readonly ContentHash Hash;
+
+        /// <summary>
+        /// Information for which <see cref="Hash"/> applies
+        /// </summary>
+        public readonly T Payload;
+
+        /// <nodoc />
+        public PayloadFromDisk(ContentHash hash, T payload)
+        {
+            Hash = hash;
+            Payload = payload;
+        }
+
+        /// <nodoc />
+        public int CompareTo(PayloadFromDisk<T> other)
+        {
+            return Hash.CompareTo(other.Hash);
+        }
+    }
+
+    /// <summary>
+    /// This class represents an immutable snapshot of a collection of items at an undetermined point in time, used for enumerations of the file
+    /// system. It is made specifically to avoid issues with enumerations of large amounts of files which go over the maximum object size
+    /// restriction.
+    /// </summary>
+    /// <typeparam name="T">Type held inside the snapshot</typeparam>
+    public class ContentHashAddressableSnapshot<T> : IEnumerable<PayloadFromDisk<T>>
+    {
+        private List<PayloadFromDisk<T>>[] _snapshot;
+
+        /// <nodoc />
+        public readonly long Count = 0;
+
+        /// <nodoc />
+        public ContentHashAddressableSnapshot()

```

Comment:
This type is immutable. Does it make sense to have a separate constructor for creating an empty instance?

Reply:
This is convenient when testing

---
In reply to: [285318012](https://github.com/microsoft/BuildXL/pull/329#discussion_r285318012) [](ancestors = 285318012)

Reply:
I disagree with this approach: in tests you could have a helper that will pass an empty enumerable.

---
In reply to: [285320330](https://github.com/microsoft/BuildXL/pull/329#discussion_r285320330) [](ancestors = 285320330,285318012)